src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -1343,19 +1343,19 @@ void LoggingEventProc()
                     }
                     else
                     {
-                        _emptyQueueEvent.Set();
+                        _emptyQueueEvent?.Set();
 
                         // Wait for next event, or finish.
-                        if (!completeAdding.IsCancellationRequested && _eventQueue.IsEmpty)
+                        if (!completeAdding.IsCancellationRequested && _eventQueue?.IsEmpty == true)
                         {
                             WaitHandle.WaitAny(waitHandlesForNextEvent);
                         }
 
-                        _emptyQueueEvent.Reset();
+                        _emptyQueueEvent?.Reset();
                     }
-                } while (!_eventQueue.IsEmpty || !completeAdding.IsCancellationRequested);
+                } while (!completeAdding.IsCancellationRequested || _eventQueue?.IsEmpty == false);
```

Comment:
The logic change here is that we do not crash if `_eventQueue` is null. Otherwise the end result logic is same. So this is good (we do not want unhandled exceptions as part of our contract)

As from perf point of view - checking both is quick - so I do not expect any measurable impact there either.

Reply:
@JanKrivanek I forgot to mention that the second part of logic update was in the timing when we stop logging based on the request. 
and the main question here was should we wait till we process all events or should we stop once we received the cancellation request :)